File: 파이썬/leetcode/dp/Longest Increasing Subsequence/Main2.py
class Solution(object):
    def lengthOfLIS(self, nums):

        def bisect(sub,target):
            left = 0
            right = len(sub)-1
            while left<=right:
                mid = (left+right)//2
                if sub[mid]<target:
                    left = mid +1
                else:
                    right = mid-1
            return left

        sub = [nums[0]]
        for i in range(1,len(nums)):
            if sub[-1]>=nums[i]:
                idx = bisect(sub,nums[i])
                sub[idx]=nums[i]
            else:
                sub.append(nums[i])
        print(sub)
        return len(sub)

# 복습

File: Programmers/src/바킹독/이분탐색/수찾기/Main2.java
package 바킹독.이분탐색.수찾기;

import java.io.*;
import java.util.*;

public class Main2 {
    public static void main(String[] args) throws IOException {
        new Main2().solution();
    }

    void solution() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::valueOf).toArray();
        int M = Integer.parseInt(br.readLine());
        int[] nums = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::valueOf).toArray();

        Arrays.sort(arr);

        StringBuilder sb = new StringBuilder();
        for (int num : nums) {
            if (bs(arr, num)) {
                sb.append("1\n");
            } else {
                sb.append("0\n");
            }
        }
        System.out.println(sb);
    }

    boolean bs(int[] arr, int num) {
        int left = 0, right = arr.length - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (arr[mid] < num) {
                left = mid + 1;
            } else if (arr[mid] == num) {
                return true;
            } else {
                right = mid - 1;
            }
        }
        return false;
    }
}
